tests/oracle.rs

```diff
@@ -0,0 +1,372 @@
+        // Add the oracle account's component as a library to link
+        // against so we can reference the account in the transaction script.
+        assembler
+            .with_library(ORACLE_COMPONENT_LIBRARY.as_ref())
+            .expect("adding oracle library should not fail")
+            .clone(),
+    )
+    .unwrap();
+    let txn_args = TransactionArgs::with_tx_script(tx_script);
+
+    let executed_transaction = executor
+        .execute_transaction(oracle_account.id(), block_ref, &[], txn_args)
+        .unwrap();
+
+    oracle_account
+        .apply_delta(executed_transaction.account_delta())
+        .unwrap();
+
+    // check that the oracle account has successfully been updated with the correct values
+    assert_eq!(
+        oracle_account.storage().slots()[1].value(),
+        oracle_data_word_1
+    );
+    assert_eq!(
+        oracle_account.storage().slots()[2].value(),
+        oracle_data_word_2
+    );
+    assert_eq!(
+        oracle_account.storage().slots()[3].value(),
+        oracle_data_word_3
+    );
+    assert_eq!(
+        oracle_account.storage().slots()[4].value(),
+        oracle_data_word_4
+    );
+}
+
+#[test]
+fn test_oracle_read() {
+    //  SETUP
+    // --------------------------------------------------------------------------------------------
+    let (oracle_pub_key, _) = get_new_pk_and_authenticator();
+    let oracle_account_id = AccountId::try_from(10376293541461622847_u64).unwrap();
+
+    let [oracle_data_1, oracle_data_2, oracle_data_3, oracle_data_4] = mock_oracle_data();
+
+    let oracle_data_word_1 = oracle_data_1.to_word();
+    let oracle_data_word_2 = oracle_data_2.to_word();
+    let oracle_data_word_3 = oracle_data_3.to_word();
+    let oracle_data_word_4 = oracle_data_4.to_word();
+
+    let oracle_storage_slots = vec![
+        StorageSlot::Value(oracle_data_word_1),
+        StorageSlot::Value(oracle_data_word_2),
+        StorageSlot::Value(oracle_data_word_3),
+        StorageSlot::Value(oracle_data_word_4),
+    ];
+
+    let oracle_account =
+        get_oracle_account(oracle_pub_key, oracle_account_id, oracle_storage_slots);
+
+    let (native_account, _) = AccountBuilder::new()
+        .init_seed(ChaCha20Rng::from_entropy().gen())
+        .with_component(
+            AccountMockComponent::new_with_slots(TransactionKernel::testing_assembler(), vec![])
+                .unwrap(),
+        )
+        .nonce(ONE)
+        .build_testing()
+        .unwrap();
+
+    let mut mock_chain =
+        MockChain::with_accounts(&[native_account.clone(), oracle_account.clone()]);
+
+    mock_chain.seal_block(None);
+
+    let advice_inputs = get_mock_fpi_adv_inputs(&oracle_account, &mock_chain);
+
+    let code = format!(
+        "
+        use.std::sys
+
+        use.miden::tx
+
+        begin
+            ### get oracle data 1 ###
+
+            # pad the stack for the `execute_foreign_procedure`execution
+            padw padw padw push.0.0
+            # => [pad(14)]
```

Comment:
Why do we need to pad 14 stack elements? Is this always the case for 
`execute_foreign_procedure`? Can you explain a bit more in a comment

Reply:
The reason is because we can call a foreign procedure with up to 15 inputs. In this case the `foreign_procedure_hash` being the required function argument (which fn do we want to exec) and then `foreign_account_id` the only input that we pass to the executed procedure.

In this example hence we have 2 elements to pass, the rest needs to be padded to make the stack 16 elements:

https://github.com/0xPolygonMiden/miden-base/blob/2b8d3dbe97305b5c2db75d64aa032fc8a675f382/miden-lib/asm/miden/tx.masm#L164-L206